cognify/optimizer/trace/checkpoint.py

```diff
@@ -0,0 +1,409 @@
+            self.best_score = score
+        if self.lowest_cost is None or price < self.lowest_cost:
+            self.lowest_cost = price
+        if self.fastest_exec_time is None or exec_time < self.fastest_exec_time:
+            self.fastest_exec_time = exec_time
+
+        logger.debug(
+            f"- {self.instance_id} - Trial id {trial_id} result: score= {score:.2f}, cost@1000= ${price*1000:.3f}, exec_time= {result.reduced_exec_time:.2f}s"
+        )
+    
+    def load_existing_logs(self) -> dict[str, TrialLog]:
+        """Load optimization trace for a BO instance
+        """
+        with open(self.opt_log_path, "r") as f:
+            opt_trace = json.load(f)
+        for trial_log_id, trial_meta in opt_trace.items():
+            trial_log = TrialLog.from_dict(trial_meta)
+            self.opt_logs[trial_log_id] = trial_log
+            self.opt_cost += trial_log.result.total_eval_cost
+            
+        cancidates = self.filter_by_constraint()
+        if cancidates:
+            self.best_score = max([log.result.reduced_score for log in cancidates])
+            self.lowest_cost = min([log.result.reduced_price for log in cancidates])
+            self.fastest_exec_time = min([log.result.reduced_exec_time for log in cancidates])
+        self._loaded = True
+        return self.opt_logs
+    
+    def save_opt_logs(self):
+        """Save the optimization trace to a json file
+        """
+        opt_logs_json_obj = {}
+        for k, v in self.opt_logs.items():
+            if v.result and v.result.complete:
+                opt_logs_json_obj[k] = v.to_dict()
+        if not opt_logs_json_obj:
+            logger.warning("No finished trials to save")
+            return
+        json.dump(opt_logs_json_obj, open(self.opt_log_path, "w+"), indent=4)
+    
+    def get_opt_summary(self):
+        """Get the pareto frontier of the layer
+        """
+        candidates = self.filter_by_constraint()
+        pareto_frontier = get_pareto_front(candidates)
+        return pareto_frontier
+
+    @property
+    def all_finished(self):
+        return all([log.result and log.result.complete for log in self.opt_logs.values()])
+    
+    def get_completed_logs(self) -> list[TrialLog]:
+        return [log for log in self.opt_logs.values() if log.result and log.result.complete]
+    
+    def filter_by_constraint(self) -> list[TrialLog]:
+        """Get logs that are qualified
+        """
+        cancidates = []
+        for log_id, log in self.opt_logs.items():
+            if not log.result or not log.result.complete:
+                continue
+            # if not meet the quality constraint, skip
+            if (
+                CommonStats.quality_constraint is not None
+                and log.result.reduced_score < CommonStats.quality_constraint
+            ):
+                continue
+            cancidates.append(log)
+        return cancidates
+
+    @property
+    def best_metrics_so_far(self):
```

Comment:
We are returning these 3-tuples everywhere, but this is error-prone. We should just return a small data structure `Metrics` that has score, cost, and time as fields. Otherwise, we have to rely on the calling code being correct.

Reply:
I will just leave this as it is in this PR, will need to refactor to customized objective later

Reply:
Created an issue to track this #50